repair/row_level.cc

```diff
@@ -1995,7 +2022,7 @@ future<> repair_init_messaging_service_handler(repair_service& rs, distributed<d
             auto from = cinfo.retrieve_auxiliary<gms::inet_address>("baddr");
             return smp::submit_to(src_cpu_id % smp::count, [from, repair_meta_id, set_diff = std::move(set_diff), needs_all_rows] () mutable {
                 auto rm = repair_meta::get_repair_meta(from, repair_meta_id);
-                return rm->get_row_diff_handler(set_diff, repair_meta::needs_all_rows_t(needs_all_rows));
+                return rm->get_row_diff_handler(std::move(set_diff), repair_meta::needs_all_rows_t(needs_all_rows));
             });
```

Comment:
This will cause a lot of cross-cpu frees, because the moved set_diff will be freed on the target cpu, not the source.

Maybe we should wrap set_diff in a foreign_ptr.

Reply:
Do you mean something like this:
```
--- a/repair/row_level.cc
+++ b/repair/row_level.cc
@@ -1760,9 +1760,10 @@ static future<stop_iteration> repair_get_row_diff_with_rpc_stream_process_op(
             return make_exception_future<stop_iteration>(std::runtime_error("get_row_diff_with_rpc_stream: Inject error in handler loop"));
         }
         bool needs_all_rows = hash_cmd.cmd == repair_stream_cmd::needs_all_rows;
-        return smp::submit_to(src_cpu_id % smp::count, [from, repair_meta_id, needs_all_rows, set_diff = std::move(current_set_diff)] {
+        auto fp = make_foreign(std::make_unique<std::unordered_set<repair_hash>>(std::move(current_set_diff)));
+        return smp::submit_to(src_cpu_id % smp::count, [from, repair_meta_id, needs_all_rows, fp = std::move(fp)] {
             auto rm = repair_meta::get_repair_meta(from, repair_meta_id);
-            return rm->get_row_diff_handler(std::move(set_diff), repair_meta::needs_all_rows_t(needs_all_rows));
+            return rm->get_row_diff_handler(std::move(*fp), repair_meta::needs_all_rows_t(needs_all_rows));
         }).then([sink] (repair_rows_on_wire rows_on_wire) mutable {
             if (rows_on_wire.empty()) {
                 return sink(repair_row_on_wire_with_cmd{repair_stream_cmd::end_of_current_rows, repair_row_on_wire()});

```

Reply:
@avikivity

Reply:
If you move *fp then you get the same delete-on-wrong-shard problem. foreign_ptr only takes care of data it owns.

Reply:
Added "repair: Wrap with foreign_ptr to avoid cross cpu free" in new version.